wremnants/theory_tools.py

```diff
@@ -207,60 +204,70 @@ def define_pdf_columns(df, dataset_name, pdfs, noAltUnc):
+
+def define_theory_weights_and_corrs(df, dataset_name, helpers, args):
+    #TODO: Reimplement this
+    #if args.highptscales:
+    #    weight_expr = f"{weight_expr}*MEParamWeightAltSet3[0]"
     df = define_prefsr_vars(df)
     df = define_ew_vars(df)
 
-    if args.theoryCorr and dataset_name in helpers:
-        helper = helpers[dataset_name]
-        df = define_theory_corr(df, weight_expr, helper, generators=args.theoryCorr, 
-                modify_central_weight=not args.theoryCorrAltOnly)
-    else:
-        df = df.Define("nominal_weight", weight_expr)
+    df = df.DefinePerSample("theory_weight_truncate", "10.")
+    df = define_central_pdf_weight(df, dataset_name, args.pdfs[0])
+    df = define_theory_corr(df, dataset_name, helpers, generators=args.theoryCorr, 
+            modify_central_weight=not args.theoryCorrAltOnly)
+    df = df.Define("nominal_weight", "exp_weight*central_pdf_weight*theory_corr_weight")
+    df = define_pdf_columns(df, dataset_name, args.pdfs, args.altPdfOnlyCentral)
+        
     return df 
 
-def define_theory_corr(df, weight_expr, helpers, generators, modify_central_weight):
+def define_theory_corr(df, dataset_name, helpers, generators, modify_central_weight):
+    df = df.Define(f"nominal_weight_uncorr", "exp_weight*central_pdf_weight")
+
+    sample_helpers = helpers[dataset_name] if dataset_name in helpers else None
+
+    if not modify_central_weight or not sample_helpers or generators[0] not in sample_helpers:
+        df = df.DefinePerSample("theory_corr_weight", "1.0")
+        return df
+
     for i, generator in enumerate(generators):
-        if i == 0:
-            if modify_central_weight and generator in helpers:
-                df = df.Define("nominal_weight_uncorr", weight_expr)
-            else:
-                df = df.Define("nominal_weight", weight_expr)
-                df = df.Alias("nominal_weight_uncorr", "nominal_weight")
-        if generator not in helpers:
+        if generator not in sample_helpers:
             continue
 
-        helper = helpers[generator]
+        helper = sample_helpers[generator]
 
         if "Helicity" in generator:
             df = df.Define(f"{generator}Weight_tensor", helper, ["massVgen", "absYVgen", "ptVgen", "chargeVgen", "csSineCosThetaPhi", "nominal_weight_uncorr"])
         elif 'ew' in generator:
             df = df.DefinePerSample(f"{generator}Dummy", "0.")
-            df = df.Define(f"{generator}Weight_tensor", helper, ["ewMll", "ewLogDeltaM", f"{generator}Dummy", "chargeVgen", "nominal_weight_uncorr" if i == 0 else "nominal_weight"]) # multiplying with nominal QCD weight
+            multiplicative_weight = i != 0 and modify_central_weight
+            df = df.Define(f"{generator}Weight_tensor", helper, ["ewMll", "ewLogDeltaM", f"{generator}Dummy", "chargeVgen", "nominal_weight_uncorr" if not multiplicative_weight else "nominal_weight"]) # multiplying with nominal QCD weight
```

Comment:
Does it work to use "nominal_weight" here? Because it is only defined after "define_theory_corr" is called (in line 236).

Reply:
Good question maybe not. Will test.